### ISSUE TITLE
Unit Testing in OSPOS

### ISSUE THREAD

**Issue by objecttothis** [opensourcepos/opensourcepos#3235]
### Background information

- [x] Reporting an issue of an unmodified OSPOS installation
- [x] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [x] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [x] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [x] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [x] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [x] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 
- OSPOS git commit hash is: 
- PHP version is: 7.4.20
- MySQL or MariaDB version is: 8.0.25
- OS and version is: FreeBSD 12.1
- WebServer is: Apache 2.4.48
- Selected language is: English
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: 
- xDebug: 3.0.4
- PhpStorm: 2021.1.4

### Question

Goal: Get unit tests to run in PhpStorm or through the CLI on the server.

@jekkos has been pushing for us to all write unit tests for functions to help prevent regressions in our code.  I wholeheartedly agree, but have been having a horrible time trying to get unit tests functioning. So, here is where I will post questions.  Perhaps this issue can be used as the basis of a wiki document on setting up unit testing.

I'm using this article (https://matthewsetter.com/php-debugging-with-phpstorm-xdebug-phpunit/) to setup

**Comment by objecttothis** [comment]
In order to setup PHPUnit in PhpStorm I've gone to File > Settings > PHP > Test Frameworks and added a new profile.  I've chosen "Use Composer autoloader".  The problem is that now I need to set the Path to script, which in a built version of OSPOS would be under /vendor/autoload.php I believe.  This means that I've encountered my first road block.  I need to build OSPOS.

**Comment by objecttothis** [comment]
PhpStorm even tells me that the vendor directory is missing and asks if I want to run 'composer install', but then gives an error:
![image](https://user-images.githubusercontent.com/17935339/124594230-9d32fd00-de70-11eb-943d-fedba777612d.png)

@jekkos @daN4cat any ideas.  composer.json exists in the \git\opensourcepos\ directory, so I'm not sure what's preventing it from being found.

**Comment by jekkos** [comment]
@objecttothis maybe we can setup a call or sth so we can debug this faster?

**Comment by objecttothis** [comment]
@jekkos thank you.  I created a discussion in the team area so that my information was not publicly visible.

**Comment by objecttothis** [comment]
@jekkos I'm on zoom now, but have not heard from you.  I'll remain on for a bit longer.  Otherwise, just respond to the discussion I created in the team area.

**Comment by objecttothis** [comment]
I was able to get composer to run in PhpStorm and I think I have all the necessary components installed, but I'm not convinced that my PHPUnit configuration is correct.  I decided for simplicity to setup WAMPServer locally and use PHP there rather than rely on my remote server... just in case that was causing the problem.  When I attempt to run a unit test I get the following output:

![image](https://user-images.githubusercontent.com/17935339/125287854-7a03c400-e32e-11eb-8e25-feef32ff8a90.png)


Is phpunit-9.5.6 compatible?  Here's my configuration... I'm not sure that's correct:

![image](https://user-images.githubusercontent.com/17935339/125288204-e7175980-e32e-11eb-9f33-cf0debc55134.png)

**Comment by objecttothis** [comment]
I saw #3196, so I tried 7.5.6 and 7.5.20, but got the same results.  @jekkos @daN4cat @SteveIreland any idea's what is missing from my configuration?

**Comment by jekkos** [comment]
Hi indeed this version of phpunit is very new, I don't think it will work. Can you check which version you have in the composer.json?

I think it should be 7.x or something. PhpUnit 9 is only compatible with PHP 8.

**Comment by objecttothis** [comment]
Yeah, composer.json was 7.5.6.  I tried that and 7.5.20 which is the latest version of composer in 7.x.  I got the same errors though.  Do I need to specify a "default configuration file?  I'm not sure my settings are correct.

**Comment by jekkos** [comment]
Can you try to use alternative bootstrap file? I think this should be the composer bootstrap file that is in the vendor folder.

Also the configuration file for phpunit is in the test folder I believe, so those two boxes should be ticked.
